src/mds/FSMap.cc

```diff
@@ -0,0 +1,762 @@
+    compat = migrate_fs->mds_map.compat;
+    enable_multiple = false;
+  } else {
+    ::decode(epoch, p);
+    ::decode(next_filesystem_id, p);
+    ::decode(legacy_client_fscid, p);
+    ::decode(compat, p);
+    ::decode(enable_multiple, p);
+    std::vector<Filesystem> fs_list;
+    ::decode(fs_list, p);
+    filesystems.clear();
+    for (std::vector<Filesystem>::const_iterator fs = fs_list.begin(); fs != fs_list.end(); ++fs) {
+      filesystems[fs->fscid] = std::make_shared<Filesystem>(*fs);
+    }
+
+    ::decode(mds_roles, p);
+    ::decode(standby_daemons, p);
+    ::decode(standby_epochs, p);
+  }
+
+  DECODE_FINISH(p);
+}
+
+
+void Filesystem::encode(bufferlist& bl) const
+{
+  ENCODE_START(1, 1, bl);
+  ::encode(fscid, bl);
+  bufferlist mdsmap_bl;
+  mds_map.encode(mdsmap_bl, CEPH_FEATURE_PGID64 | CEPH_FEATURE_MDSENC);
+  ::encode(mdsmap_bl, bl);
+  ENCODE_FINISH(bl);
+}
+
+void Filesystem::decode(bufferlist::iterator& p)
+{
+  DECODE_START(1, p);
+  ::decode(fscid, p);
+  bufferlist mdsmap_bl;
+  ::decode(mdsmap_bl, p);
+  bufferlist::iterator mdsmap_bl_iter = mdsmap_bl.begin();
+  mds_map.decode(mdsmap_bl_iter);
+  DECODE_FINISH(p);
+}
+
+int FSMap::parse_filesystem(
+      std::string const &ns_str,
+      std::shared_ptr<Filesystem> *result
+      ) const
+{
+  std::string ns_err;
+  fs_cluster_id_t fscid = strict_strtol(ns_str.c_str(), 10, &ns_err);
+  if (!ns_err.empty() || filesystems.count(fscid) == 0) {
+    for (auto fs : filesystems) {
+      if (fs.second->mds_map.fs_name == ns_str) {
+        *result = fs.second;
+        return 0;
+      }
+    }
+    return -ENOENT;
+  } else {
+    *result = get_filesystem(fscid);
+    return 0;
+  }
+}
+
+void Filesystem::print(std::ostream &out) const
+{
+  // TODO add a non-json print?
+  JSONFormatter f;
+  dump(&f);
+  f.flush(out);
+}
+
+mds_gid_t FSMap::find_standby_for(mds_role_t role, const std::string& name) const
+{
+  mds_gid_t result = MDS_GID_NONE;
+
+  // First see if we have a STANDBY_REPLAY
+  auto fs = get_filesystem(role.fscid);
+  for (const auto &i : fs->mds_map.mds_info) {
+    const auto &info = i.second;
+    if (info.rank == role.rank && info.state == MDSMap::STATE_STANDBY_REPLAY) {
+      return info.global_id;
+    }
+  }
+
+  // See if there are any STANDBY daemons available
+  for (const auto &i : standby_daemons) {
+    const auto &gid = i.first;
+    const auto &info = i.second;
+    assert(info.state == MDSMap::STATE_STANDBY);
+    assert(info.rank == MDS_RANK_NONE);
+
+    if (info.laggy()) {
+      continue;
+    }
+
+    if ((info.standby_for_rank == role.rank && info.standby_for_ns == role.fscid)
+        || (name.length() && info.standby_for_name == name)) {
+      // It's a named standby for *me*, use it.
+      return gid;
+    } else if (info.standby_for_rank < 0 && info.standby_for_name.length() == 0)
+      // It's not a named standby for anyone, use it if we don't find
+      // a named standby for me later.
+      result = gid;
```

Comment:
Heh, so we'll always use the last one in the list since we set result on every iteration? Sounds fine to me as long as you say it's okay.